### PR TITLE
Topic argocd e2e tests

### PR DESCRIPTION
**Pull request by preethignanesh**
With the Argo CD integrated into verrazzano, we need to add e2e tests for Argo CD verification.
By default, Argo CD is disabled in all profiles, so we are testing for Argo CD in the below 2 tests by explicitly enabling it - 
* Dynamic-update
* Private-registry

### CODE REVIEW

File: tests/e2e/pkg/argocd.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+		return err
+	}
+	var emptyList bool
+	gomega.Eventually(func() (bool, error) {
+		contains, err := GetApplicationsWithClient(log, argocdURL, token)
+		emptyList = contains
+		return emptyList, err
+	}, waitTimeout, pollingInterval).Should(gomega.BeTrue())
+
+	gomega.Expect(emptyList).To(gomega.BeTrue(), "Argocd UI is accessible and no applications are deployed")
```
Comment (anders-swanson) [inline tests/e2e/pkg/argocd.go:66]: Do we need to test deploying an application and verifying it is there?

File: tests/e2e/pkg/argocd.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	return nil
+}
+
+func eventuallyGetArgocdAdminPassword(log *zap.SugaredLogger) (string, error) {
+	var err error
+	var secret *corev1.Secret
+	gomega.Eventually(func() error {
+		secret, err = GetSecret("argocd", "argocd-initial-admin-secret")
```
Comment (anders-swanson) [inline tests/e2e/pkg/argocd.go:74]: can you use ArgoCDNamespace instead of string literal

File: tests/e2e/pkg/argocd.go (hunk 1 above)
Comment (preethignanesh) [reply tests/e2e/pkg/argocd.go:66]: Deploying an application would require us to connect to the github repo which contains some application manifests to be deployed to the cluster. Can we connect to the github repo for deploying the application using ginkgo as I am not aware of and need to check. Is it required ?

File: tests/e2e/verify-infra/restapi/argocd_url_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+var _ = t.Describe("argocd", Label("f:infra-lcm",
+	"f:ui.console"), func() {
+	t.Context("test to", func() {
+		t.It("Verify argocd access and configuration", func() {
+			kubeconfigPath, err := k8sutil.GetKubeConfigLocation()
+			if err != nil {
+				t.Logs.Error(fmt.Sprintf("Error getting kubeconfig: %v", err))
+				t.Fail(err.Error())
+			}
+			if pkg.IsArgoCDEnabled(kubeconfigPath) {
```
Comment (anders-swanson) [inline tests/e2e/verify-infra/restapi/argocd_url_test.go:25]: probably want to skip test if argocd not enabled

Comment (anders-swanson) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go:25]: see ginkgo.Skip

File: tests/e2e/verify-infra/restapi/argocd_url_test.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+				if err != nil {
+					t.Logs.Error(fmt.Sprintf("Error verifying access to Argocd: %v", err))
+					t.Fail(err.Error())
+				}
+
+				metrics.Emit(t.Metrics.With("argocd_access_elapsed_time", time.Since(start).Milliseconds()))
+
+				start = time.Now()
+				t.Logs.Info("Accessing the Argocd Applications")
+				err = pkg.VerifyArgocdApplicationAccess(t.Logs, kubeconfigPath)
```
Comment (anders-swanson) [inline tests/e2e/verify-infra/restapi/argocd_url_test.go:38]: Is this a separate spec? seems like multiple things are getting tested in the same spec

File: tests/e2e/verify-infra/restapi/argocd_url_test.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+			if pkg.IsArgoCDEnabled(kubeconfigPath) {
+
+				start := time.Now()
+				err = pkg.VerifyArgoCDAccess(t.Logs, kubeconfigPath)
+				if err != nil {
+					t.Logs.Error(fmt.Sprintf("Error verifying access to Argocd: %v", err))
+					t.Fail(err.Error())
+				}
+
+				metrics.Emit(t.Metrics.With("argocd_access_elapsed_time", time.Since(start).Milliseconds()))
```
Comment (anders-swanson) [inline tests/e2e/verify-infra/restapi/argocd_url_test.go:34]: you are emitting this metric twice

File: tests/e2e/verify-install/verrazzano/verrazzano_test.go [hunk 6]
```diff
@@ -846,3 +854,82 @@ func assertPodAntiAffinity(matchLabels map[string]string, namespace string) {
...
+		var pods []corev1.Pod
+
+		Eventually(func() bool {
+			pods, err = pkg.GetPodsFromSelector(&metav1.LabelSelector{MatchLabels: map[string]string{"app.kubernetes.io/instance": "argocd"}}, constants.ArgoCDNamespace)
+			if err != nil {
+				t.Logs.Error("Failed to get Argocd pods: %v", err)
+				return false
+			}
+			if len(pods) < 1 {
+				return false
```
Comment (anders-swanson) [inline tests/e2e/verify-install/verrazzano/verrazzano_test.go:873]: log message here

File: tests/e2e/pkg/argocd.go (hunk 1 above)
Comment (anders-swanson) [reply tests/e2e/pkg/argocd.go:66]: Any way to fake it or create a dummy git repo (not necessarily in github)

File: tests/e2e/verify-install/verrazzano/verrazzano_test.go (hunk 6 above)
Comment (preethignanesh) [reply tests/e2e/verify-install/verrazzano/verrazzano_test.go:874]: Logged the message

File: tests/e2e/verify-infra/restapi/argocd_url_test.go (hunk 3 above)
Comment (preethignanesh) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go]: Done

File: tests/e2e/pkg/argocd.go (hunk 2 above)
Comment (preethignanesh) [reply tests/e2e/pkg/argocd.go]: Done

File: tests/e2e/verify-infra/restapi/argocd_url_test.go (hunk 4 above)
Comment (preethignanesh) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go]: We are trying to check if the URL is accessible and then trying to check the empty list of applications within Argo CD

File: tests/e2e/verify-infra/restapi/argocd_url_test.go (hunk 5 above)
Comment (preethignanesh) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go]: First metric when we try to access the login page and the second metric when we access the applications page within Argo CD

**Review by anders-swanson — APPROVED** [review]

File: tests/e2e/verify-infra/restapi/argocd_url_test.go (hunk 4 above)
Comment (preethignanesh) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go]: Closing

File: tests/e2e/verify-infra/restapi/argocd_url_test.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	"f:ui.console"), func() {
+	t.Context("test to", func() {
+		t.It("Verify argocd access and configuration", func() {
+			kubeconfigPath, err := k8sutil.GetKubeConfigLocation()
+			if err != nil {
+				t.Logs.Error(fmt.Sprintf("Error getting kubeconfig: %v", err))
+				t.Fail(err.Error())
+			}
+
+			if pkg.IsArgoCDEnabled(kubeconfigPath) {
```
Comment (fredtibbitts) [inline tests/e2e/verify-infra/restapi/argocd_url_test.go:26]: This check (and the one above) are probably better done in a `BeforeEach` (helps keep the test focused on what's actually being tested), but it's minor.

File: tests/e2e/verify-infra/restapi/argocd_url_test.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+				Skip("Skipping Argo CD access test as Argo CD is not enabled.")
+			}
+
+			start := time.Now()
+			err = pkg.VerifyArgoCDAccess(t.Logs, kubeconfigPath)
```
Comment (fredtibbitts) [inline tests/e2e/verify-infra/restapi/argocd_url_test.go:31]: Looking at this test it's not clear what the difference is between `VerifyArgoCDAccess` and `VerifyArgocdApplicationAccess`. Either make the function names more accurately describe what's being accessed or add comments so it's clear.

File: tests/e2e/verify-infra/restapi/argocd_url_test.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/verrazzano/verrazzano/pkg/k8sutil"
+	"github.com/verrazzano/verrazzano/tests/e2e/pkg"
+	"github.com/verrazzano/verrazzano/tests/e2e/pkg/test/framework/metrics"
+)
+
+var _ = t.Describe("argocd", Label("f:infra-lcm",
+	"f:ui.console"), func() {
+	t.Context("test to", func() {
+		t.It("Verify argocd access and configuration", func() {
```
Comment (fredtibbitts) [inline tests/e2e/verify-infra/restapi/argocd_url_test.go:19]: The Describe/Context/It are supposed to read similar to Given/When/Then, so something like:

`Describe("ArgoCD")`
`Context("is enabled")`
`It("the web application is accessible")`

Something like that.

**Review by fredtibbitts — COMMENTED** [review]
A couple of general comments. The function/variable naming is inconsistent, it's `...ArgoCD...` in some places and `...Argocd...` in others.

I agree with @anders-swanson comment about testing by deploying a real app. I'm not exactly sure which tests/pipeline that belongs in, and it can certainly be done in another PR, but it seems like the tests are incomplete without that.

File: tests/e2e/pkg/argocd.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+	return token, nil
+}
+
+func GetApplicationsWithClient(log *zap.SugaredLogger, argoCDURL string, token string) (bool, error) {
+	argoCDLoginURL := fmt.Sprintf("%s/%s", argoCDURL, "api/v1/applications")
+	client := &http.Client{}
+	var bearer = "Bearer " + token
+
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true} /* #nosec G402 */
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/pkg/argocd.go:133]: Inorder to run the Argo CD with ingress we need to run with TLS disabled. So, when we setup the pods as well, we disable the TLS which is as per the argo docs - https://argo-cd.readthedocs.io/en/stable/operator-manual/ingress/#option-1-ssl-passthrough

File: tests/e2e/verify-infra/restapi/argocd_url_test.go (hunk 9 above)
Comment (preethignanesh) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go]: Done

File: tests/e2e/verify-infra/restapi/argocd_url_test.go (hunk 8 above)
Comment (preethignanesh) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go]: Added comments

File: tests/e2e/verify-infra/restapi/argocd_url_test.go (hunk 7 above)
Comment (preethignanesh) [reply tests/e2e/verify-infra/restapi/argocd_url_test.go]: Added BeforeEach

File: tests/e2e/verify-install/verrazzano/verrazzano_test.go [hunk 11]
```diff
@@ -846,3 +854,83 @@ func assertPodAntiAffinity(matchLabels map[string]string, namespace string) {
...
+
+// Validates the Argocd namespace and deployments if Argocd is explicitly enabled
+func validateArgoCDResources() {
+
+	var err error
+
+	if argocdEnabled {
+		var pods []corev1.Pod
+
+		Eventually(func() bool {
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/verify-install/verrazzano/verrazzano_test.go:864]: If I use ShouldNot(BeEmpty()), then the spec is failing but with ShouldNot(BeNil()) it works.

**Comment by preethignanesh** [comment]
As suggested by @anders-swanson and @fredtibbitts, have added argocd app deployment to the e2e test.

File: tests/e2e/verify-install/verrazzano/verrazzano_test.go [hunk 12]
```diff
@@ -846,3 +854,82 @@ func assertPodAntiAffinity(matchLabels map[string]string, namespace string) {
...
+			if err != nil {
+				t.Logs.Error("Failed to get Argocd pods: %v", err)
+				return false
+			}
+			if len(pods) < 1 {
+				t.Logs.Error("No pods exist in the Argo CD namespace")
+				return false
+			}
+			return true
+		}, waitTimeout, pollingInterval).ShouldNot(BeNil())
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/verify-install/verrazzano/verrazzano_test.go]: changed

File: tests/e2e/verify-install/keycloak/keycloak_test.go [hunk 13]
```diff
@@ -553,6 +568,33 @@ func verifyRancherClientURIs(keycloakClient *Client, env string) bool {
...
+		t.Logs.Error(fmt.Printf("Incorrect Number of WebOrigins returned for client %+v\n", keycloakClient.WebOrigins))
+		return false
+	}
+
+	// Verify rancher redirectUI
+	if !verifyURIs(keycloakClient.RedirectUris, argocdURI+env, 1) {
+		t.Logs.Error(fmt.Printf("Expected 1 ArgoCD redirect URIs. Found %+v\n", keycloakClient.RedirectUris))
+		return false
+	}
+	// Verify rancher web origin
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/verify-install/keycloak/keycloak_test.go]: renamed to argocd

File: tests/e2e/verify-install/keycloak/keycloak_test.go [hunk 14]
```diff
@@ -553,6 +568,33 @@ func verifyRancherClientURIs(keycloakClient *Client, env string) bool {
...
+		return false
+	}
+
+	// Verify Correct number of WebOrigins
+	if len(keycloakClient.WebOrigins) != 1 {
+		t.Logs.Error(fmt.Printf("Incorrect Number of WebOrigins returned for client %+v\n", keycloakClient.WebOrigins))
+		return false
+	}
+
+	// Verify rancher redirectUI
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/verify-install/keycloak/keycloak_test.go]: renamed to argocd

File: tests/e2e/pkg/argocd.go [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+		log.Errorf("Failed to extract token from argocd response: %v", err)
+		return false, err
+	}
+
+	exists := strings.Contains(token, "resourceVersion")
+	return exists, nil
+
+}
+
+// CreateArgoCDGitApplication Creates a Fake Git Repo
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/pkg/argocd.go]: changed the comments

File: tests/e2e/pkg/argocd.go [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+	token, err := httputil.ExtractFieldFromResponseBodyOrReturnError(string(body), "token", "unable to find token in Argocd response")
+	if err != nil {
+		log.Errorf("Failed to extra token from argocd response: %v", err)
+		return "", err
+	}
+
+	return token, nil
+}
+
+func GetApplicationsWithClient(log *zap.SugaredLogger, argoCDURL string, token string) (bool, error) {
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/pkg/argocd.go:139]: Added the comments

File: tests/e2e/examples/argocd-helidon/argocd_helidon_suite_test.go [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"flag"
+	"testing"
+
+	"github.com/onsi/ginkgo/v2"
+)
+
+var skipDeploy bool
+var skipUndeploy bool
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/examples/argocd-helidon/argocd_helidon_suite_test.go]: Removed the undeploy variable

File: tests/e2e/examples/argocd-helidon/argocd_helidon_test.go [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+func helloHelidonPodsRunning() bool {
+	result, err := pkg.PodsRunning(namespace, expectedPodsHelloHelidon)
+	if err != nil {
+		AbortSuite(fmt.Sprintf("One or more pods are not running in the namespace: %v, error: %v", namespace, err))
+	}
+	return result
+}
+
+func appEndpointAccessible(url string, hostname string) bool {
```
[the remark this answers is not in the document]
Comment (preethignanesh) [reply tests/e2e/examples/argocd-helidon/argocd_helidon_test.go:119]: Its not clear whether I have to change the return value. If so, can we get this addressed in a separate PR ?

**Review by fredtibbitts — APPROVED** [review]

**Pull request merged**